src/aspire/estimation/covar2d.py

```diff
@@ -108,21 +110,21 @@ def get_mean(self, coeffs, ctf_fb=None, ctf_idx=None):
 
         if (ctf_fb is None) or (ctf_idx is None):
             ctf_idx = np.zeros(coeffs.shape[1], dtype=int)
-            ctf_fb = [blk_diag_eye(blk_diag_partition(RadialCTFFilter().fb_mat(self.basis)))]
+            ctf_fb = [BlkDiagMatrix.eye(get_partition(RadialCTFFilter().fb_mat(self.basis)),dtype=coeffs.dtype)]
 
         b = np.zeros(self.basis.count, dtype=coeffs.dtype)
 
-        A = blk_diag_zeros(blk_diag_partition(ctf_fb[0]), dtype=coeffs.dtype)
+        A = BlkDiagMatrix.zeros(get_partition(ctf_fb[0]), dtype=coeffs.dtype)
         for k in np.unique(ctf_idx[:]).T:
             coeff_k = coeffs[:, ctf_idx == k]
             weight = np.size(coeff_k, 1)/np.size(coeffs, 1)
             mean_coeff_k = self._get_mean(coeff_k)
             ctf_fb_k = ctf_fb[k]
-            ctf_fb_k_t = blk_diag_transpose(ctf_fb_k)
-            b = b + weight*blk_diag_apply(ctf_fb_k_t, mean_coeff_k)
-            A = blk_diag_add(A, blk_diag_mult(weight, blk_diag_mult(ctf_fb_k_t, ctf_fb_k)))
+            ctf_fb_k_t = ctf_fb_k.T
+            b = b + ctf_fb_k_t.apply(mean_coeff_k) * weight
+            A += weight * (ctf_fb_k_t @ ctf_fb_k)
```

Comment:
Your concern is correct... but because `A` is created just above, the scope is local, and I think this is okay here.

Generally speaking yes, with structures of arrays, we will always need to consider whether we are mutating something we care about. There was only one place I saw a note from another developer where they did a deep copy, and I think I got that right.  I'll leave this comment open for now and come back after the minor stuff.

Reply:
Upon review, I think internally we only need to consider in-place optimization dangerous for BlkDiagMatrix arguments to methods. `b_covar` comes to mind in the cov2d code, but I think I have those correct as they are.

Probably best to consider it an optimization and use it sparingly.  It easily be de-optimized (in fact that is what python would do for in place if no method was overloaded, so it would just amount to deleting...). Maybe its not worth the risk?

Reply:
Yeah I think it makes sense. I just get burned on this periodically, coming from MATLAB, but I guess you get used to it after a while.